sumpy/e2e.py

```diff
@@ -190,19 +200,84 @@ def get_translation_loopy_insns(self):
+                complex_dtype=to_complex_dtype(result_dtype),
                 )
 
-    def get_kernel(self):
-        ncoeff_src = len(self.src_expansion)
+    def get_postprocess_loopy_insns(self, result_dtype):
+        """Loopy instructions that happen only once for each target box.
+
+        :arg result_dtype: A numpy dtype for the result. This is important
+            because depending on the input to the M2L being real or not
+            the code needs to be different. If the input was real, the
+            result should be the real part of the complex values from the
+            inverse FFT. In the input was complex, the result matches the
+            values from the inverse FFT.
+        """
+
         ncoeff_tgt = len(self.tgt_expansion)
+        if self.use_precomputed_exprs:
+            nprecomputed_exprs = \
+                self.tgt_expansion.m2l_global_precompute_nexpr(self.src_expansion)
+        else:
+            nprecomputed_exprs = 0
+
+        if self.use_fft:
+            ncoeff_tgt = nprecomputed_exprs
+
+        from sumpy.assignment_collection import SymbolicAssignmentCollection
+        sac = SymbolicAssignmentCollection()
+
+        tgt_coeff_exprs = [
+            sym.Symbol("coeff_sum%d" % i) for i in range(ncoeff_tgt)
+        ]
+
+        src_rscale = sym.Symbol("src_rscale")
+        tgt_rscale = sym.Symbol("tgt_rscale")
+
+        if self.use_fft:
+            tgt_coeff_post_exprs = self.tgt_expansion.m2l_postprocess_exprs(
+                self.src_expansion, tgt_coeff_exprs, src_rscale, tgt_rscale,
+                sac=sac)
+            if result_dtype in (np.float32, np.float64):
+                real_func = sym.Function("real")
+                tgt_coeff_post_exprs = [real_func(expr) for expr in
+                    tgt_coeff_post_exprs]
+        else:
+            tgt_coeff_post_exprs = tgt_coeff_exprs
+
+        tgt_coeff_post_names = [
+            sac.assign_unique("coeff_post%d" % i, coeff)
+            for i, coeff in enumerate(tgt_coeff_post_exprs)
+        ]
+
+        sac.run_global_cse()
 
+        from sumpy.codegen import to_loopy_insns
+        insns = to_loopy_insns(
+                sac.assignments.items(),
+                vector_names=set(["d"]),
+                pymbolic_expr_maps=[self.tgt_expansion.get_code_transformer()],
+                retain_names=tgt_coeff_post_names,
+                complex_dtype=to_complex_dtype(result_dtype),
+                )
+        return insns
+
+    def get_kernel(self, result_dtype):
         if self.use_precomputed_exprs:
             nprecomputed_exprs = \
                 self.tgt_expansion.m2l_global_precompute_nexpr(self.src_expansion)
         else:
             nprecomputed_exprs = 0
 
+        if self.use_fft:
+            ncoeff_src = nprecomputed_exprs
+            ncoeff_post = nprecomputed_exprs
+        else:
+            ncoeff_src = len(self.src_expansion)
+            ncoeff_post = len(self.tgt_expansion)
```

Comment:
?

Reply:
Clarified in d83705cb388fe230eff6801220416df36b63bdb0